File: javascript/leetcode/min-stack.js
/**
 * initialize your data structure here.
 */
var MinStack = function() {
    this.min = null
    this.collection = []
};

/** 
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function(x) {
    if (this.min == null || this.min > x) { this.min = x }
    this.collection.push(x)
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function() {
    result = this.collection.pop()
    if (result == this.min) {
        if (this.collection.length == 0) {
            this.min = null
        } else {
            this.min = Infinity
            for (var i = 0; i < this.collection.length; i++) {
                if (this.min > this.collection[i]) {this.min = this.collection[i]}
            }
        }
    }
    return result
};

/**
 * @return {number}
 */
MinStack.prototype.top = function() {
    return this.collection[this.collection.length - 1]
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function() {
    return this.min
};

/** 
 * Your MinStack object will be instantiated and called as such:
 * var obj = Object.create(MinStack).createNew()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: java/leetcode/min-stack.java
class MinStack {
    static int min;
    static ArrayList<Integer> collection;
    /** initialize your data structure here. */
    public MinStack() {
        min = Integer.MAX_VALUE;
        collection = new ArrayList<Integer>();
    }
    
    public void push(int x) {
        if (min > x) { min = x; }
        collection.add(x);
    }
    
    public void pop() {
        int result = collection.remove(collection.size() - 1);
        if (result == min) {
            min = Integer.MAX_VALUE;
            for (int i = 0; i < collection.size(); i++) {
                if (min > collection.get(i)) {min = collection.get(i);}
            }
        }
    }
    
    public int top() {
        return collection.get(collection.size() - 1);
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
 